doc/design.md

```diff
@@ -0,0 +1,249 @@
+until the Event Handler returns the status as complete.
+
+Covers:
+
+* `req~iterating-over-loop~1`
+
+Needs: impl, utest, itest
+
+
+
+### Returning JSON Result
+`dsn~returning-json-result~1`
+
+The Event Loop returns result in json format.
+
+Covers:
+
+* `req~returning-result~1`
+
+Needs: impl, utest, itest
+
+
+
+## Event Handler
+The Event Handler is a framework whose logic is implemented by users. It reacts 
+to the query results of  the Event Loop and generates SQL queries of the next 
+state. It stores the states in BucketFS.
+
+
+### Python UDF Framework 
+`dsn~python-udf-framework~1`
+
+An Event Handler designed as Python UDF  allows users to have an easy-to-use and 
+better tooling framework to build their algorithm on top it. The user can develop 
+the algorithm as a Python UDF script in which the framework is imported. Some 
+details of this framework interface are listed below:
+
+- The event handler is implemented as a class and states are kept as attributes 
+of a class to facilitate the states operations.
+- The user code should implement a method `handle_event` where the logic of the 
+code is placed.
+- The event handler is initiated by getting the following inputs: The results of 
+the return query,  the next state information including its parameters, 
+context object. 
+- Upon completion of the algorithm, the Event Handler calls methods itself to 
+remove temporary records.
+- In case of an error caught in the Event Loop, the cleanup event is called for 
+the Event Handler and the temporary records are deleted.
+
+Covers:
+
+* `req~user-friendly-framework-interface~1`
+
+Needs: impl, utest, itest
+
+
+### Storing States in BucketFS 
+`dsn~storing-states-in-bucketfs~1`
+
+The framework keeps states during iterations by storing them in BucketFS.
+
+Covers:
+
+* `req~managing-temporary-bucketfs-files~1`
+
+Needs: impl, utest, itest
+
+
+It returns the sequence of SQL queries to execute
+A query to call the same UDF to run the action for the next state
+
+### Returning Queries
+`dsn~returning-queries~1`
+
+The Event Handler returns a list of SQL queries to execute and the return query 
+which will be called again as the next state action so that the new  state 
+transition is performed.
+
+Covers:
+
+* `feat~implementation-framework~1`
+
+Needs: impl, utest, itest
+
+### Managing Temporary BucketFS Files
+`dsn~managing-temporary-bucketfs-files~1`
+
+The Event Handler create temporary BucketFS files that will be  kept as result. 
+These temporary files are placed  in the same directory in  the BucketFS.   
+
+Covers:
+
+* `req~managing-temporary-bucketfs-files~1`
+
+Needs: impl, utest, itest
+
+
+### Managing Temporary BucketFS Files
+`dsn~managing-temporary-tables~1`
+
+The Event Handler uses temporary tables to store large intermediate results. 
+
+Covers:
+
+* `req~managing-temporary-tables~1`
+
+Needs: impl, utest, itest
+
+
+### Electing Leader
+`dsn~electing-leader~1`
+
```

Comment:
Before leader election can happen. UDFs need to discover each other. We will use pycos, but it works, in general, the following. Every UDF receives on a fixed port UDP packages and sends UDP broadcasts on the same port. This way, everyone should find everyone. Assuming the Network has enough capacity left to send the UDP packages. To connect not to UDF instances of a different UDF call, each UDF call includes a unique ID which will be part of the broadcast message. We use this ID to connect only to other UDF instances of the same UDF Call.

Further, we know how many UDF instances should have been started, and we wait with leader election, until all instances have connected to each other.